lib/compress/zstd_fast.c

```diff
@@ -613,75 +633,200 @@ static size_t ZSTD_compressBlock_fast_extDict_generic(
     if (prefixStartIndex == dictStartIndex)
         return ZSTD_compressBlock_fast(ms, seqStore, rep, src, srcSize);
 
-    /* Search Loop */
-    while (ip < ilimit) {  /* < instead of <=, because (ip+1) */
-        const size_t h = ZSTD_hashPtr(ip, hlog, mls);
-        const U32    matchIndex = hashTable[h];
-        const BYTE* const matchBase = matchIndex < prefixStartIndex ? dictBase : base;
-        const BYTE*  match = matchBase + matchIndex;
-        const U32    curr = (U32)(ip-base);
-        const U32    repIndex = curr + 1 - offset_1;
+    /* start each op */
+_start: /* Requires: ip0 */
+
+    step = stepSize;
+    nextStep = ip0 + kStepIncr;
+
+    /* calculate positions, ip0 - anchor == 0, so we skip step calc */
+    ip1 = ip0 + 1;
+    ip2 = ip0 + step;
+    ip3 = ip2 + 1;
+
+    if (ip3 >= ilimit) {
+        goto _cleanup;
+    }
+
+    hash0 = ZSTD_hashPtr(ip0, hlog, mls);
+    hash1 = ZSTD_hashPtr(ip1, hlog, mls);
+
+    idx = hashTable[hash0];
+    idxBase = idx < prefixStartIndex ? dictBase : base;
+
+    do {
+        U32 mval; /* src or dict value at match idx */
+
+        /* load repcode match for ip[2] */
+        const U32 current2 = (U32)(ip2 - base);
+        const U32 repIndex = current2 - offset_1;
         const BYTE* const repBase = repIndex < prefixStartIndex ? dictBase : base;
-        const BYTE* const repMatch = repBase + repIndex;
-        hashTable[h] = curr;   /* update hash table */
-        DEBUGLOG(7, "offset_1 = %u , curr = %u", offset_1, curr);
-
-        if ( ( ((U32)((prefixStartIndex-1) - repIndex) >= 3) /* intentional underflow */
-             & (offset_1 <= curr+1 - dictStartIndex) ) /* note: we are searching at curr+1 */
-           && (MEM_read32(repMatch) == MEM_read32(ip+1)) ) {
-            const BYTE* const repMatchEnd = repIndex < prefixStartIndex ? dictEnd : iend;
-            size_t const rLength = ZSTD_count_2segments(ip+1 +4, repMatch +4, iend, repMatchEnd, prefixStart) + 4;
-            ip++;
-            ZSTD_storeSeq(seqStore, (size_t)(ip-anchor), anchor, iend, REPCODE1_TO_OFFBASE, rLength);
-            ip += rLength;
-            anchor = ip;
+        U32 rval;
+        assert(offset_1 > 0);
+        if ( ( ((U32)(prefixStartIndex - repIndex) >= 4) /* intentional underflow */
+               & (offset_1 < current2 - dictStartIndex) ) ) {
```

Comment:
Note: `offset_1 == current2 - dictStartIndex` is surprisingly possible in streaming compression, since the offset can be carried over from a previous noDict block. The fuzzers actually found a case where this happens. We need strict inequality to satisfy an assertion below.

More broadly, this predicate should be pulled into a helper function and documented (along with the same predicate in all other extDict / DMS matchfinders). I will do that in June or July. @terrelln

Reply:
It seems this condition is `unlikely()` to be false ?

Reply:
My understanding is that this test went from `<=` to `<` because there is an unconditional look-back-by-one-byte later on. This should probably be mentioned. Alternatively, the "unconditional" nature of the "look-back-by-one-byte" operation should be revisited.

Reply:
Also, presuming that the `offset_1 == current2 - dictStartIndex` can be taken care of,
I wonder if it is even possible for `offset_1 > current2 - dictStartIndex`,
and if it is not, should it be replaced by an `assert()`?

Reply:
I think `offset_1 > current2 - dictStartIndex` is possible in streaming compression if using a sufficiently large dictionary. A very large offset could be carried over into a future extDict block which would have a smaller dictionary (because the "dictionary" in the future block is really just part of the window).

However, this case (along with the case `offset_1 == current2 - dictStartIndex`) could be addressed by checking the offset at the top of the extDict function, before entering the search loop. Once in the search loop, it's impossible to revert to that state. I'll try this in addition to marking the remaining condition UNLIKELY and see if there are any gains. If so, I can carry over this optimization to the other dictionary matchfinders.

Reply:
> I think offset_1 > current2 - dictStartIndex is possible in streaming compression if using a sufficiently large dictionary

Yes, that's a very good point.
And the following optimization suggestion is very good too.